Some bugfixes & new features for docstubgen

seal331 opened the pull request:
This PR does the following for docstubgen:

- gets rid of a redundant error message which already has a better alternative in the code;
- fixes a corner case involving an extra binary alone on a line;
- implements recursive searching;
- implements a proper "no binary found" error condition;
- redirects all errors to stderr;
- adds binary checking support;
- moves docstubgen to a subfolder in the source tree;
- adds a manpage;
- probably at least one other thing I forgot :)

I plan on at least attempting to implement another tool for the contributor-tools section, but I need to take a break, I've been doing too much work on infra.

seal331 committed:
docstubgen: improve an error message

The "error: nonexistent folder1" message dates back to the
very first version of the script committed. Now that we have
proper error handling for nonexistent folders in a different
place (commit 6a3aa5) with an error message that actually
tells you which folder is nonexistent, this check is both
redundant and less descriptive.

seal331 committed:
docstubgen: rework folder list handling

seal331 committed:
docstubgen: fix corner case involving single binaries on new line

seal331 committed:
docstubgen: implement recursive searching

seal331 committed:
docstubgen: implement "no binary found" error conditions and redirect errors to stderr

seal331 committed:
docstubgen: allow disabling warnings

seal331 committed:
docstubgen: add binary documenting support

seal331 committed:
docstubgen: add binary check security check bypass

ONLY use this as a last resort on the most trusted programs you package.

seal331 committed:
docstubgen: add support for overriding binary detect warning without env variable

seal331 committed:
docstubgen: move to contributor-tools subfolder

seal331 committed:
contributor-tools: make sure we always get the newest version

seal331 committed:
docstubgen: final touches

seal331 committed:
gitignore: ignore contributor tools object folder

seal331 committed:
docstubgen: add compat symlink

We don't break existing workflows here.

Zeckmathederg commented:
Will take a look at this when I have working internet again, besides phone data.

Zeckmathederg reviewed (changes requested):
You should add contributor-tools/ to `CONTRIBUTING.md` in the internal section, mention what it's for.

seal331 committed:
CONTRIBUTING: document contributor-tools; change generate-doc-stub.py description

seal331 committed:
CONTRIBUTING: oops, forgot to also put this here

seal331 commented:
> You should add contributor-tools/ to `CONTRIBUTING.md` in the internal section, mention what it's for.

Fixed in the new commits, also fixed a misleading "also documents directories" saying in the description as docstubgen does not in fact do that, and that's not coming anytime soon (I can't think of literally any way to distinguish standard system folders from custom ones without making an awfully long list of all standard system directories which isn't good coding practice anyways).

Zeckmathederg reviewed (approved):
LGTM, merging. Thanks!
